src/mame/misc/vampire.cpp

```diff
@@ -0,0 +1,644 @@
+#include "machine/pit8253.h"
+
+#include "machine/timer.h"
+#include "sound/dac.h"
+
+#include "emupal.h"
+#include "screen.h"
+#include "speaker.h"
+
+
+namespace {
+
+class vampire_state : public driver_device
+{
+public:
+	vampire_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_audiocpu(*this,"audiocpu"),
+		m_palette(*this, "palette"),
+		m_gfxrom(*this, "gfx")
+	{ }
+
+	void vampire(machine_config &config);
+	void init_vampire();
+
+protected:
+	virtual void machine_start() override;
+
+private:
+	required_device<cpu_device> m_maincpu;
+	required_device<cpu_device> m_audiocpu;
+	required_device<palette_device> m_palette;
+
+	void palette(palette_device &palette) const;
+
+	uint8_t blitter_r(offs_t offset);
+	uint8_t fsoundlatch_r(offs_t offset);
+	void audioack_w(offs_t offset, uint8_t state);
+	void blitter_w(offs_t offset, uint8_t data);
+	void blit_trigger(uint8_t * blitterdata, uint16_t data);
+
+	struct
+	{
+		int base_x = 0;
+		int base_y = 0;
+
+		int last_offset = 0;
+		int last_data = 0;
+
+		int flags = 0xff;
+
+		std::unique_ptr<uint8_t[]> layer_1;
+		std::unique_ptr<uint8_t[]> layer_2;
+		std::unique_ptr<uint8_t[]> slots;
+	} m_blitter;
+		
+	int m_audio_nmi = 0;
+	int m_audio_latch = 0;
+
+	int m_ppi_output = 0;
+
+	void vampire_memory(address_map &map);
+	void vampire_audio(address_map &map);
+
+	void pit_out_w0(int state);
+	void pit_out_w1(int state);
+	void pit_out_w2(int state);
+
+	uint8_t sound_sync_r(offs_t offset);
+	void soundlatch_w(offs_t offset, uint8_t data);
+	uint8_t sound_ack_r(offs_t offset);
+	void blitter_flags_w(offs_t offset, uint8_t data);
+	void blitter_control_w(offs_t offset, uint8_t data);
+	uint8_t io814_r(offs_t offset);
+	uint8_t io815_r(offs_t offset);
+
+	uint32_t screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+
+	required_region_ptr<uint8_t> m_gfxrom;
+};
+
+void vampire_state::machine_start()
+{
+	m_blitter.layer_1 = make_unique_clear<uint8_t[]>(256 * 256);
+	m_blitter.layer_2 = make_unique_clear<uint8_t[]>(256 * 256);
+	m_blitter.slots = make_unique_clear<uint8_t[]>(0x1000);
+
+	save_item(NAME(m_blitter.base_x));
+	save_item(NAME(m_blitter.base_y));
+	save_item(NAME(m_blitter.last_offset));
+	save_item(NAME(m_blitter.last_data));
+	save_item(NAME(m_blitter.flags));
+	save_pointer(NAME(m_blitter.layer_1), 256 * 256);
+	save_pointer(NAME(m_blitter.layer_2), 256 * 256);
+	save_pointer(NAME(m_blitter.slots), 0x1000);
+
+	save_item(NAME(m_audio_nmi));
+	save_item(NAME(m_audio_latch));
+
+	save_item(NAME(m_ppi_output));
+}
+
+uint8_t vampire_state::fsoundlatch_r(offs_t offset)
+{
+	return m_audio_latch;
+}
+
+void vampire_state::soundlatch_w(offs_t offset, uint8_t data)
+{
+	m_audio_nmi = 0x00;
+	m_audio_latch = data;
+	m_audiocpu->set_input_line(INPUT_LINE_NMI, ASSERT_LINE);	
+	machine().scheduler().perfect_quantum(attotime::from_usec(10));
+}
```

Comment:
If you don’t use `synchronize(...)` to set `m_audio_latch`, you’ll still get potential issues because the sound CPU will see the latch value change before it “catches up” and sees the NMI